lib/package_source.py

```diff
@@ -0,0 +1,162 @@
+from subprocess import Popen, PIPE
+import os
+import shutil
+import urllib2
+
+
+from lib import config
+from lib import repository
+from lib import utils
+
+
+def _hg_download(source, directory):
+    """
+    Clones a mercurial [source] to [directory] and returns a source dict.
+    """
+    CONF = config.get_config().CONF
+    proxy = CONF.get('http_proxy')
+    hg_source = source['hg']
+    repo_name = os.path.basename(hg_source['src'])
+    dest = os.path.join(directory, repo_name)
+    command = 'hg '
+
+    if proxy:
+        command += '-c http_proxy.host="{}" '.format(proxy)
+
+    commit_id = hg_source.get('commit_id')
+    branch = hg_source.get('branch')
+
+    if commit_id is None and branch is None:
+        raise ValueError('invalid hg source dict: missing both `commit_id` '
+                         'and `branch`')
+
+    command += 'clone "{}" -r "{}" "{}"'.format(hg_source['src'],
+                                                hg_source['branch'],
+                                                dest)
+    proc = Popen(command.split(' '), stderr=PIPE)
+    proc.communicate()
```

Comment:
It is preferable use lib.utils.run_command to execute external commands instead of use directly Popen.